File: julia/problem_12.jl
# Copyright 2014 Mitchell Kember. Subject to the MIT License.
# Project Euler: Problem 12
# Highly divisible triangular number

module problem_12

# off by one for perfect squares
function divisors(n)
	count = 2
	for i = 2:isqrt(n)
		if n % i == 0
			count += 2
		end
	end
	count
end

function main()
	tri, i = 1, 1
	while true
		if divisors(tri) > 500
			return tri
		end
		i += 1
		tri += i
	end
end

end # module

File: rust/problem_12.rs
// Copyright 2014 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 12
// Highly divisible triangular number

// Returns the number of divisors that n has. The result is incorrect for
// perfect squares (off by one).
fn divisors(n: int) -> int {
	let mut count = 2;
	let max = (n as f32).sqrt() as int + 1;
	for i in range(2, max) {
		if n % i == 0 {
			count += 1;
		}
	}
	count * 2
}

pub fn solve() -> int {
	let mut tri = 1;
	let mut i = 1;
	loop {
		if divisors(tri) > 500 {
			return tri;
		}
		i += 1;
		tri += i;
	}
}
